### PR TITLE
EVG-2415 Auto switch between spot and on-demand

### PR DESCRIPTION
**Pull request by bsamek**

### CODE REVIEW

File: cloud/ec2_price.go [hunk 1]
```diff
@@ -24,16 +25,28 @@ type timeRange struct {
...
+	pkgCachingPriceFetcher.Lock()
+	return pkgCachingPriceFetcher
+}
+
+func putPkgCachingPriceFetcher(fetcher *cachingPriceFetcher) {
+	defer fetcher.Unlock()
+	if fetcher != pkgCachingPriceFetcher {
+		grip.EmergencyPanic("fetcher passed to putPkgCachingPriceFetcher is not package fetcher")
 	}
+}
```
Comment (tychoish) [inline cloud/ec2_price.go]: I worry that this is going to be pretty unexpected. Let's work on getting an alternate approach to this together.

File: cloud/ec2_price.go [hunk 2]
```diff
@@ -128,9 +139,73 @@ func (cpf *cachingPriceFetcher) cacheEc2Prices() error {
...
+		}
+	}
+	return min, nil
+}
+
+func (m *ec2Manager) getProvider(h *host.Host, ec2settings *NewEC2ProviderSettings) (ec2ProviderType, error) {
+	if m.provider == spotProvider {
+		return spotProvider, nil
+	}
+	if m.provider == onDemandProvider {
```
Comment (tychoish) [inline cloud/ec2_price.go]: shouldn't these consts' be in the global package, or use them?

File: cloud/ec2_price.go (hunk 1 above)
Comment (bsamek) [reply cloud/ec2_price.go]: Sure.

File: cloud/ec2_price.go (hunk 2 above)
Comment (bsamek) [reply cloud/ec2_price.go]: No, these are specific to this implementation, are not used anywhere else, and are an iota.

File: cloud/ec2_price.go (hunk 1 above)
Comment (bsamek) [reply cloud/ec2_price.go]: I pushed a different implementation that doesn't use the get/put indirection.

File: cloud/ec2_price.go [hunk 3]
```diff
@@ -30,21 +30,6 @@ type cachingPriceFetcher struct {
...
-	pkgCachingPriceFetcher.Lock()
-	return pkgCachingPriceFetcher
-}
-
-func putPkgCachingPriceFetcher(fetcher *cachingPriceFetcher) {
-	defer fetcher.Unlock()
-	if fetcher != pkgCachingPriceFetcher {
-		grip.EmergencyPanic("fetcher passed to putPkgCachingPriceFetcher is not package fetcher")
-	}
-}
```
Comment (tychoish) [inline cloud/ec2_price.go]: I think having a get method is fine, as long as you do the new in a package init()

I didn't like the panic and the put, but I think the get is fine

File: cloud/new_ec2_cost_test.go [hunk 4]
```diff
@@ -275,41 +275,42 @@ func (s *CostIntegrationSuite) TestGetProviderAuto() {
 	settings := &NewEC2ProviderSettings{}
 	s.m.provider = autoProvider
 
-	fetcher := getPkgCachingPriceFetcher()
+	if pkgCachingPriceFetcher == nil {
+		pkgCachingPriceFetcher = new(cachingPriceFetcher)
+	}
+	pkgCachingPriceFetcher.Lock()
+	defer pkgCachingPriceFetcher.Unlock()
```
Comment (tychoish) [inline cloud/new_ec2_cost_test.go]: would it be more reasonable to write this as a suite and not use the global instance of the fetcher?

File: cloud/ec2_price.go (hunk 3 above)
Comment (bsamek) [reply cloud/ec2_price.go]: I don't think they're necessary at all, actually, after moving to an init function. I simplified the logic.

File: cloud/new_ec2_cost_test.go (hunk 4 above)
Comment (bsamek) [reply cloud/new_ec2_cost_test.go]: There's no need for locking in the test at all. I've removed it.

**Comment by bsamek** [comment]
I moved the locks.

**Pull request merged**